netket/vqs/mc/mc_state/state.py

```diff
@@ -293,38 +294,50 @@ def sampler(self, sampler: Sampler):
                 )
             )
 
+        # Save the old `n_samples` before the new `sampler` is set.
+        # `_chain_length == 0` means that this `MCState` is being constructed.
+        if self._chain_length > 0:
+            n_samples_old = self.n_samples
+
         self._sampler = sampler
         self.sampler_state = self.sampler.init_state(
             self.model, self.variables, seed=self._sampler_seed
         )
+
+        # Update `n_samples`, `n_samples_per_rank`, and `chain_length` according
+        # to the new `sampler.n_chains`.
+        # If `n_samples` is divisible by the new `sampler.n_chains`, it will be
+        # unchanged; otherwise it will be rounded up.
+        # If the new `n_samples_per_rank` is not divisible by `chunk_size`, a
+        # `ValueError` will be raised.
+        # `_chain_length == 0` means that this `MCState` is being constructed.
+        if self._chain_length > 0:
+            self.n_samples = n_samples_old
```

Comment:
```suggestion
        if self._chain_length > 0:
            self.n_samples = self.n_samples
```
doesn't this work?

Reply:
No, this will not work. We need to first compute the old `n_samples` using the old `sampler.n_chains` and save it, then assign the new sampler, then set `self.n_samples` and compute the new `chain_length` using the new `sampler.n_chains`. That's really a bit tricky...

Reply:
Hmm..